File: CodeForces/347_Div2/Rebus.py
import sys
import psyco
sys.stdin = open("/home/shiva/Learning/1.txt", "r")
sys.stdout = open("/home/shiva/Learning/2.txt", "w")

def compute(plus,minus,total,inp):
	if plus == 1 and minus == 0:
		print(total); return
	elif (plus == 1 and minus == 1): 
		print("Impossible"); return
	elif (abs(plus-minus) > total):
		plus

	temp = total
	total += minus
	res = []
	if int(total/plus) > temp:
		print("Impossible"); return
	elif int(total%plus) == 0:
		res = [int(total/plus) for i in range(0,plus)]
	else:
		res = [int(total/(plus-1)) for i in range(0,plus-1)]
		res.append(total%(plus-1))
	
	j = 0
	prev = 0
	for i in inp.split():
		if j == 0:
			print(res[j],end=' ')
			j+=1
		elif i == '+' or i=='-':
			print(i,end=' ')
			prev = i
		elif i == '?':
			if prev == '+':

				print(res[j],end=' ')
				j+=1
			else:
				print('1',end=' ')
		else:
			print(i,end=' ')

inp = input()
plus =1
minus = 0
total = 0
for i in inp.split():
	if i=='?' or i=='=':
		continue
	elif i == '+':
		plus+=1
	elif i == '-':
		minus +=1
	else:
		total = int(i)

compute(plus,minus,total,inp)

File: CodeForces/347_Div2/Rebus.cpp
#include <bits/stdc++.h>
#define vi vector<int>
#define pi pair<int,int>
#define vii vector<pi>
#define ll long long
using namespace std;
int main() {
#ifndef ONLINE_JUDGE
	freopen("/home/shiva/Learning/1.txt", "r", stdin);
	freopen("/home/shiva/Learning/2.txt", "w", stdout);
#endif
	ios::sync_with_stdio(false); cin.tie(0);
	char c; int n;
	bool arr[110]; memset(arr, 0, sizeof arr);
	int val[110];
	arr[0] = true;
	int cnt = 1, plus = 1, minus = 0;
	while (cin >> c) {
		if (c == '=') break;
		if (c == '+') {arr[cnt++] = true; plus++;}
		else if (c == '-') { arr[cnt++] = false; minus++;}
	}
	for (int i = 0; i < (plus + minus); i++) val[i] = 1;
	cin >> n;
	int add = n - plus + minus;
	bool pos = (add >= 0);
	if (add < 0) add = -add;
	for (int i = 0; i < cnt; i++) {
		if (arr[i] == pos) {
			int diff = n - val[i];
			diff = min(diff, add);
			val[i] += diff;
			add -= diff;
		}
	}
	bool possible = (add == 0);
	if (possible) {
		cout << "Possible" << endl;
		cout << val[0];
		for (int i = 1; i < cnt; i++) {
			if (arr[i]) cout << " + ";
			else cout << " - ";
			cout << val[i];
		}
		cout << " = " << n << endl;

	} else cout << "Impossible" << endl;

}
